pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,567 @@
+			activeServices[serviceName] = true
+			serviceIP := net.ParseIP(service.Spec.ClusterIP)
+			info, exists := proxier.getServiceInfo(serviceName)
+			if exists && proxier.sameConfig(info, service, servicePort) {
+				// Nothing changed.
+				continue
+			}
+			if exists {
+				//Something changed.
+				glog.V(4).Infof("Something changed for service %q: stopping it", serviceName)
+				proxier.removeService(serviceName, info)
+			}
+
+			glog.V(1).Infof("Adding new service %q at %s:%d/%s", serviceName, serviceIP, servicePort.Port, servicePort.Protocol)
+			info = proxier.addService(serviceName, servicePort.Protocol)
+			info.portal.ip = serviceIP
+			info.portal.port = servicePort.Port
+			info.deprecatedPublicIPs = service.Spec.DeprecatedPublicIPs
+			// Deep-copy in case the service instance changes
+			info.loadBalancerStatus = *api.LoadBalancerStatusDeepCopy(&service.Status.LoadBalancer)
+			info.sessionAffinityType = service.Spec.SessionAffinity
+
+			glog.V(4).Infof("info: %+v", info)
+		}
+	}
+
+	func() {
+		proxier.mu.Lock()
+		defer proxier.mu.Unlock()
+		for name, info := range proxier.serviceMap {
+			if !activeServices[name] {
+				glog.V(1).Infof("Removing service %q", name)
+				proxier.removeServiceInternal(name, info)
+			}
+		}
+	}()
+
+	glog.V(6).Infof("Syncing rules.")
+	if err := proxier.iptablesInit(); err != nil {
+		glog.Errorf("Failed to ensure iptables: %v", err)
+	}
+	// this locks proxier.mu, hence the closure above
+	if err := proxier.syncProxyRules(); err != nil {
+		glog.Errorf("Failed to sync iptables rules: %v", err)
+	}
+}
+
+// OnEndpointsUpdate takes in a slice of updated endpoints.
+func (proxier *Proxier) OnEndpointsUpdate(allEndpoints []api.Endpoints) {
+	registeredEndpoints := make(map[proxy.ServicePortName]bool)
+	proxier.mu.Lock()
+	defer proxier.mu.Unlock()
+
+	// Update endpoints for services.
+	for i := range allEndpoints {
+		svcEndpoints := &allEndpoints[i]
+
+		// We need to build a map of portname -> all ip:ports, protocol for that
+		// portname.  Explode Endpoints.Subsets[*] into this structure.
+		portsToEndpoints := map[string][]hostPortProtocol{}
+		for i := range svcEndpoints.Subsets {
+			ss := &svcEndpoints.Subsets[i]
+			for i := range ss.Ports {
+				port := &ss.Ports[i]
+				for i := range ss.Addresses {
+					addr := &ss.Addresses[i]
+					portsToEndpoints[port.Name] = append(portsToEndpoints[port.Name], hostPortProtocol{addr.IP, port.Port, port.Protocol})
+				}
+			}
+		}
+
+		for portname := range portsToEndpoints {
+			svcPort := proxy.ServicePortName{types.NamespacedName{svcEndpoints.Namespace, svcEndpoints.Name}, portname}
+			state, exists := proxier.serviceMap[svcPort]
+			if !exists || state == nil {
+				// This should be safe, a ServicePortName should map to at least one port on a service and should have one protocol
+				state = proxier.addService(svcPort, portsToEndpoints[portname][0].protocol)
```

Comment:
added an `addServiceInternal` call which does not hit the mutex. checking for other deadlocks.